tests/test_proto.py

```diff
@@ -0,0 +1,10 @@
+def test_imports():
+    from pysubstrait.proto.pysubstrait.algebra_pb2 import Expression
```

Comment:
Just curious about the `pysubstrait.proto.pysubstrait...` could we make it something like
`from pysubstrait.proto.core.algebra_pb2 import Expression` or just `from pysubstrait.proto.algebra_pb2 import Expression`

Reply:
I would also have preferred that, but the second `pysubstrait` was added as a namespace for the generated protobuf classes so as not to conflict with other implementations like C++. I used this script in `gen_proto.sh` to generate it: https://github.com/substrait-io/substrait/blob/main/tools/proto_prefix.py. I'm open to a better naming convention if anyone has one!

Reply:
Aha, I understand your point. :thinking:

Reply:
We can always merge as-is and change it in the future while the project is still pre- v1.0. For reference, Ibis-substrait sort of does it this way, too: https://github.com/ibis-project/ibis-substrait/tree/main/ibis_substrait/proto/substrait/ibis

Reply:
Actually "standard" (as in, the message defining the hint is in the core substrait proto files) hints will probably be more of a danger than extensions.

Reply:
I think we haven't actually run into this in ibis-substrait yet, but I don't know that anyone has tried it.

Reply:
Do you mind elaborating on the dangers with an example?

Reply:
Sure.  Let's pretend I created a message (`EstimatedSelectivity`) that was intended to be used as an advanced extension on `RelCommon::Hint` and we merged it into Substrait:

```
message RelCommon {
  ...
  Hint hint = 3;
  ...
  message Hint {
    ...
    substrait.extensions.AdvancedExtension advanced_extension = 10;
    ...
  }
  ...
}

message EstimatedSelectivity {
  double selectivity = 1;
}
```

Then, I attached that hint onto a filter relation.  Here is the plan, serialized as JSON:

```
{
  ...
  "filter": {
    "common": {
      "hint": {
        "optimization": {
          "@type": "substrait.EstimatedSelectivity",
          "selectivity": 0.03
        }
      }
    },
    "condition": { ... },
    "input": { ... }
  }
  ...
}
```

As you can see, the Substrait package is included as part of the `@type` when an `Any` is serialized.

However, I don't know of any existing use like this.  In Acero we use `google.protobuf.Any` to add new nodes (currently as-of join and segmented aggregation).  In both cases the new message is not part of the core Substrait set of messages.  It's a part of Acero's custom protobuf.  This custom protobuf uses Substrait messages (e.g. an as-of join has field references in it and these are Substrait expressions) but the only type name that gets serialized is the extension itself.

I think it's reasonable to expect that will never be the case.  For example, consider the above example.  One could argue that a new "standard hint" should be included in the hint definition as:

```
message RelCommon {
  ...
  Hint hint = 3;
  ...
  message Hint {
    ...
    EstimatedSelectivity estimated_selectivity = 11;
    substrait.extensions.AdvancedExtension advanced_extension = 10;
    ...
  }
  ...
}
```

In other words, any time `google.protobuf.Any` is used it must refer to an object that is not part of the core Substrait protobuf files.

However, I'm not completely convinced it will never happen.

Reply:
I believe what Weston is describing is how I planned to approach doing my own extensions. I think the most concise example is here, where I use ibis-substrait to translate and then I use a `google.protobuf.Any` for an `ExtensionLeafRel`:
https://github.com/drin/mohair-extension/blob/mainline/mohair_extension/extension.py#L100

Using just the standard string representation, here's what I generate:
https://media.githubusercontent.com/media/drin/mohair-extension/mainline/resources/projection.mohair.txt


> an object that is not part of the core Substrait protobuf files

I am assuming that if something becomes core, then I would change the way this is done. Using ibis-substrait's approach to translation makes it an easy change, not sure that approach holds true for any other, large projects.

Reply:
Thank you, both! The examples are super helpful. It's a good thing to note!